storage/shard.go

```diff
@@ -1543,6 +1572,75 @@ func (s *dbShard) Flush(
 	return s.markFlushStateSuccessOrError(blockStart, multiErr.FinalError())
 }
 
+func (s *dbShard) Snapshot(
+	blockStart time.Time,
+	snapshotTime time.Time,
+	flush persist.Flush,
+) error {
+	// We don't snapshot data when the shard is still bootstrapping
+	s.RLock()
+	if s.bs != bootstrapped {
+		s.RUnlock()
+		return errShardNotBootstrappedToSnapshot
+	}
+	s.RUnlock()
+
+	var multiErr xerrors.MultiError
+
+	s.markIsSnapshotting()
+	defer func() {
+		s.markDoneSnapshotting(multiErr.Empty(), snapshotTime)
+	}()
+
+	prepareOpts := persist.PrepareOptions{
+		NamespaceMetadata: s.namespace,
+		Shard:             s.ID(),
+		BlockStart:        blockStart,
+		SnapshotTime:      snapshotTime,
+		FilesetType:       persist.FilesetSnapshotType,
+	}
+	prepared, err := flush.Prepare(prepareOpts)
+	// Add the err so the defer will capture it
+	multiErr = multiErr.Add(err)
+	if err != nil {
+		return err
+	}
+
+	// No action is necessary therefore we bail out early and there is no need to close.
+	if prepared.Persist == nil {
+		errMsg := "[invariant violated] tried to write prepare snapshot file that already exists"
+		// This should never happen in practice since we don't check for duplicate snapshot files
+		// in the Prepare method.
+		s.logger.WithFields(
+			xlog.NewField("blockStart", blockStart),
+			xlog.NewField("snapshotStart", snapshotTime),
+			xlog.NewField("shard", s.ID()),
+		).Errorf(errMsg)
+		// Add to multiErr so we can see failures in metrics
+		multiErr = multiErr.Add(errors.New(errMsg))
+		return nil
+	}
+
+	tmpCtx := context.NewContext()
+	s.forEachShardEntry(func(entry *dbShardEntry) bool {
+		series := entry.series
+		// Use a temporary context here so the stream readers can be returned to
+		// pool after we finish fetching flushing the series
+		tmpCtx.Reset()
+		err := series.Snapshot(tmpCtx, blockStart, prepared.Persist)
+		tmpCtx.BlockingClose()
+		multiErr = multiErr.Add(err)
+		// If we encounter an error when persisting a series, we continue regardless.
+		return true
```

Comment:
Is that a good idea?  The volume will likely be bad/unuseable if there was a write error.

Reply:
you're right this is probably a bad idea, refactoring to return false

Reply:
Actually I did this because thats what the flush method does. Thoughts on remaining consistent or changing both?